kotlinx-coroutines-core/common/src/internal/SegmentList.kt

```diff
@@ -0,0 +1,213 @@
+ * Essentially, this is a node in the Michael-Scott queue algorithm, but with
+ * maintaining [prev] pointer for efficient [remove] implementation.
+ */
+internal abstract class Segment<S : Segment<S>>(val id: Long, prev: S?, pointers: Int) {
+    // Pointer to the next segment, updates similarly to the Michael-Scott queue algorithm.
+    private val _next = atomic<Any?>(null)
+    val nextOrClosed: NextSegmentOrClosed<S> get() = NextSegmentOrClosed(_next.value)
+    fun trySetNext(value: S): Boolean = _next.compareAndSet(null, value)
+
+    // Pointer to the previous segment, updates in [remove] function.
+    val prev = atomic(prev)
+
+    /**
+     * Cleans the pointer to the previous segment.
+     */
+    fun cleanPrev() = prev.lazySet(null)
+
+    /**
+     * This property should return the maximal number of slots in this segment,
+     * it is used to define whether the segment is logically removed.
+     */
+    abstract val maxSlots: Int
+
+    // numbers of cleaned slots (lowest bits) and AtomicRef pointers to this segment (highest bits)
+    private val cleanedAndPointers = atomic(pointers shl POINTERS_SHIFT)
+
+    /**
+     * Returns `true` if this segment is logically removed from the queue.
+     * The segment is considered as removed if all the slots are cleaned,
+     * there is no pointers to this segment from outside, and
+     * it is not a physical tail in the linked list of segments.
+     */
+    inline val removed get() = cleanedAndPointers.value == maxSlots && _next.value !== null
+
+    // increments the number of pointers if this segment is not logically removed\
+    inline fun tryIncPointers() = cleanedAndPointers.addConditionally(1 shl POINTERS_SHIFT) { it != maxSlots }
+
+    // returns `true` if this segment is logically removed after the decrement
+    inline fun decPointers() = cleanedAndPointers.addAndGet(-(1 shl POINTERS_SHIFT)) == maxSlots
+
+    /**
+     * This functions should be invoked on each slot clean-up;
+     * should not be invoked twice for the same slot.
+     */
+    fun onSlotCleaned() {
+        if (cleanedAndPointers.incrementAndGet() == maxSlots) remove()
+    }
+
+    /**
+     * Tries to mark the linked list as closed by forbidding adding new segments after this one.
+     */
+    fun markAsClosed() = _next.compareAndSet(null, CLOSED)
+
+    /**
+     * Checks whether this segment is a physical tail and is closed for further segment additions.
+     */
+    val isClosed = _next.value === CLOSED
+
+    /**
+     * Removes this segment physically from the segment queue. The segment should be
+     * logically removed (so [removed] returns `true`) at the point of invocation.
+     */
+    fun remove() {
+        assert { removed } // The segment should be logically removed at first
+        // Read `next` and `prev` pointers.
+        var next = this.nextOrClosed.segment ?: return // tail cannot be removed
```

Comment:
Since we move pointers using only `findSegmentAndMoveForward` function, we guarantee that there is always a pointer which references to the last segment, so that  `cleanedAndPointers != 0` for the last segment and we do not invoke `remove` on it.